drivers/spi/spi_silabs_eusart.c

```diff
@@ -199,35 +591,144 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
 	if (spi_context_rx_buf_on(&data->ctx)) {
 		UNALIGNED_PUT(rx_frame, (uint8_t *)data->ctx.rx_buf);
 	}
+
 	spi_context_update_rx(&data->ctx, 1, 1);
+
 	return 0;
 }
 
-static void spi_silabs_eusart_xfer(const struct device *dev, const struct spi_config *config)
+static int spi_silabs_eusart_xfer_dma(const struct device *dev, const struct spi_config *config)
 {
-	int ret;
+#ifdef CONFIG_SPI_SILABS_EUSART_DMA
+	const struct spi_silabs_eusart_config *eusart_config = dev->config;
 	struct spi_silabs_eusart_data *data = dev->data;
 	struct spi_context *ctx = &data->ctx;
+	int ret;
+
+	size_t total_padded_transaction_size = spi_silabs_longest_transfer_size(ctx);
+
+	spi_silabs_eusart_clear_txrx_fifos(eusart_config->base);
+
+	ret = spi_silabs_prepare_dma_transaction(dev, total_padded_transaction_size);
+
+	if (ret) {
+		return ret;
+	}
+
+	spi_silabs_eusart_pm_policy_get(dev);
+
+	spi_context_cs_control(ctx, true);
+
+	if (total_padded_transaction_size == 0U) {
+		goto force_transaction_close;
+	}
+
+	/* RX channel needs to be ready before TX channel actually starts */
+	ret = dma_start(data->dma_channel_rx.dma_dev, data->dma_channel_rx.channel_nb);
+
+	if (ret) {
+		dma_stop(data->dma_channel_tx.dma_dev, data->dma_channel_tx.channel_nb);
```

Comment:
Did you mean to put this here? I would think it should go in the next `if (ret)` below, and stop RX, not TX.